### PR TITLE
Feature/c exception codes

### PR DESCRIPTION
**Pull request by Pheubel**
Wraps code that can throw within a try-catch block to return error codes.

### CODE REVIEW

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 1]
```diff
@@ -116,7 +166,13 @@ NrtResult Nrt_AudioService_setMusicVolume(NrtAudioService service, float value)
     }
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->setMusicVolume(value);
+
+    try { 
+      serv->setMusicVolume(value);
+    } catch( const Exceptions::NotInitialisedException){ // todo: handle error message
+
```
Comment (capnkenny) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:173]: We don't return an error here?

File: src/NovelRT.Interop/DotNet/NrtRuntimeService.cpp [hunk 2]
```diff
@@ -36,7 +36,17 @@ NrtResult Nrt_RuntimeService_initialise(NrtRuntimeService service) {
...
 
     NovelRT::DotNet::RuntimeService* cService = reinterpret_cast<NovelRT::DotNet::RuntimeService*>(service);
-    cService->initialise();
+
+    try {
+      cService->initialise();
+    } catch (const Exceptions::InitialisationFailureException) {
+      Nrt_setErrMsgIsInitialisationFailureInternal();
+      return NRT_FAILURE_INITIALISATION_FAILURE;
+    } catch (const Exceptions::FunctionNotFoundException) {
```
Comment (capnkenny) [inline src/NovelRT.Interop/DotNet/NrtRuntimeService.cpp:45]: Nit: @RubyNova do we want the styling to be maintained like this? (Where each catch is inlined with bracing?)

File: src/NovelRT.Interop/Maths/NrtGeoBounds.cpp [hunk 3]
```diff
@@ -50,29 +50,21 @@ extern "C" {
       return NRT_FAILURE_NULLPTR_PROVIDED;
     }
 
-    try {
+
```
Comment (capnkenny) [inline src/NovelRT.Interop/Maths/NrtGeoBounds.cpp:53]: Nit: extra space :)

File: src/NovelRT.Interop/Animation/NrtSpriteAnimatorState.cpp [hunk 4]
```diff
@@ -124,7 +124,14 @@ NrtResult Nrt_SpriteAnimatorFrameVector_getFrameAtIndex(NrtSpriteAnimatorFrameVe
 
     std::vector<Animation::SpriteAnimatorFrame> cppVector = *reinterpret_cast<std::vector<Animation::SpriteAnimatorFrame>*>(vector);
     Animation::SpriteAnimatorFrame* cppFrame = new Animation::SpriteAnimatorFrame();
-    *cppFrame = cppVector.at(index);
+
```
Comment (capnkenny) [inline src/NovelRT.Interop/Animation/NrtSpriteAnimatorState.cpp:127]: Nit: extra space :)

File: include/NovelRT.Interop/NrtLogLevel.h [hunk 5]
```diff
@@ -13,7 +13,7 @@ typedef enum {
     NOVELRT_LOGLEVEL_INFO = 2,
     NOVELRT_LOGLEVEL_WARN = 3,
     NOVELRT_LOGLEVEL_ERR  = 4,
-    NOVELRT_LOGLEVE_OFF = 6
```
Comment (capnkenny) [inline include/NovelRT.Interop/NrtLogLevel.h:16]: Thanks for fixing my mistake lol :smile:

**Review by capnkenny — CHANGES_REQUESTED** [review]
Well done! A couple questions so I can't approve _just_ yet - I'll wait for @RubyNova to confirm the styling issue I saw, and if you could review the other comments, that'd be awesome!

File: src/NovelRT.Interop/DotNet/NrtRuntimeService.cpp (hunk 2 above)
Comment (RubyNova) [reply src/NovelRT.Interop/DotNet/NrtRuntimeService.cpp:45]: We should move to allman when you feel is appropriate.

File: include/NovelRT.Interop/NrtInteropErrorHandlingInternal.h [hunk 6]
```diff
@@ -11,6 +11,9 @@ void Nrt_setErrMsgIsNullptrInternal();
 void Nrt_setErrMsgIsNaNInternal();
 void Nrt_setErrMsgIsDivideByZeroInternal();
 void Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
+void Nrt_setErrMsgIsNotSupportedInternal();
+void Nrt_setErrMsgIsInitialisationFailureInternal();
+void Nrt_setErrMsgIsFunctionNotFound();
```
Comment (RubyNova) [inline include/NovelRT.Interop/NrtInteropErrorHandlingInternal.h:16]: Please mark function as internal.

File: src/NovelRT.Interop/Animation/NrtSpriteAnimatorState.cpp [hunk 7]
```diff
@@ -124,7 +124,16 @@ NrtResult Nrt_SpriteAnimatorFrameVector_getFrameAtIndex(NrtSpriteAnimatorFrameVe
...
+    try
+    {
+      *cppFrame = cppVector.at(index);
+    }
+    catch (const std::out_of_range)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Animation/NrtSpriteAnimatorState.cpp:133]: Was this TODO handled or are we leaving it for this PR?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 8]
```diff
@@ -50,7 +69,16 @@ NrtResult Nrt_AudioService_setSoundVolume(NrtAudioService service, unsigned int
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->setSoundVolume(source, val);
+
+    try
+    {
+      serv->setSoundVolume(source, val);
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:78]: Was this TODO handled or are we leaving it for this PR?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 9]
```diff
@@ -61,7 +89,16 @@ NrtResult Nrt_AudioService_setSoundPosition(NrtAudioService service, unsigned in
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->setSoundPosition(source, posX, posY);
+
+    try
+    {
+      serv->setSoundPosition(source, posX, posY);
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:98]: Was this TODO handled or are we leaving it for this PR?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 10]
```diff
@@ -72,7 +109,16 @@ NrtResult Nrt_AudioService_resumeMusic(NrtAudioService service) {
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->resumeMusic();
+
+    try
+    {
+      serv->resumeMusic();
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:118]: Was this TODO handled or are we leaving it for this PR?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 11]
```diff
@@ -83,7 +129,16 @@ NrtResult Nrt_AudioService_playMusic(NrtAudioService service, NrtAudioServiceIte
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->playMusic(reinterpret_cast<NovelRT::Audio::SoundBank::iterator&>(handle), loops);
+
+    try
+    {
+      serv->playMusic(reinterpret_cast<NovelRT::Audio::SoundBank::iterator&>(handle), loops);
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:138]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 12]
```diff
@@ -94,7 +149,16 @@ NrtResult Nrt_AudioService_pauseMusic(NrtAudioService service) {
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->pauseMusic();
+
+    try
+    {
+      serv->pauseMusic();
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:158]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 13]
```diff
@@ -105,7 +169,16 @@ NrtResult Nrt_AudioService_stopMusic(NrtAudioService service) {
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->stopMusic();
+
+    try
+    {
+      serv->stopMusic();
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:178]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 14]
```diff
@@ -116,7 +189,16 @@ NrtResult Nrt_AudioService_setMusicVolume(NrtAudioService service, float value)
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->setMusicVolume(value);
+
+    try
+    {
+      serv->setMusicVolume(value);
+    }
+    catch( const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:198]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 15]
```diff
@@ -138,7 +221,16 @@ NrtResult Nrt_AudioService_loadSound(NrtAudioService service, char* input, unsig
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    *output = serv->loadSound(input);
+
+    try
+    {
+      *output = serv->loadSound(input);
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:230]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Audio/NrtAudioService.cpp [hunk 16]
```diff
@@ -160,7 +252,16 @@ NrtResult Nrt_AudioService_playSound(NrtAudioService service, unsigned int handl
...
 
     auto serv = reinterpret_cast<Audio::AudioService*>(service);
-    serv->playSound(handle, loops);
+
+    try
+    {
+      serv->playSound(handle, loops);
+    }
+    catch (const Exceptions::NotInitialisedException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Audio/NrtAudioService.cpp:261]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Graphics/NrtFontSet.cpp [hunk 17]
```diff
@@ -21,7 +21,15 @@ extern "C" {
...
 
     FontSet* fontSetPtr = reinterpret_cast<FontSet*>(fontSet);
-    fontSetPtr->loadFontAsTextureSet(std::string(file), fontSize);
+
+    try
+    {
+      fontSetPtr->loadFontAsTextureSet(std::string(file), fontSize);
+    }
+    catch (const Exceptions::InitialisationFailureException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Graphics/NrtFontSet.cpp:30]: This looks like a custom exception. Do we have an internal function for the error message? Also the code being used here doesn't match the exception being caught?

File: src/NovelRT.Interop/Graphics/NrtTextRect.cpp [hunk 18]
```diff
@@ -135,7 +135,14 @@ extern "C" {
     }
 
     TextRect* textRectPtr = reinterpret_cast<TextRect*>(rect);
-    textRectPtr->setText(std::string(inputText));
+    try
+    {
+      textRectPtr->setText(std::string(inputText));
+    }
+    catch (const Exceptions::CharacterNotFoundException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Graphics/NrtTextRect.cpp:143]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Graphics/NrtTexture.cpp [hunk 19]
```diff
@@ -20,7 +20,15 @@ extern "C" {
...
 
     Texture* texturePtr = reinterpret_cast<Texture*>(targetTexture);
-    texturePtr->loadPngAsTexture(std::string(file));
+
+    try
+    {
+      texturePtr->loadPngAsTexture(std::string(file));
+    }
+    catch (const Exceptions::InvalidOperationException)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Graphics/NrtTexture.cpp:29]: This looks like a custom exception. Do we have an internal function for the error message?

File: src/NovelRT.Interop/Maths/NrtQuadTree.cpp [hunk 20]
```diff
@@ -146,7 +156,14 @@ extern "C" {
       return NRT_FAILURE_NULLPTR_PROVIDED;
     }
 
-    *outputPoint = reinterpret_cast<NrtQuadTreePoint>(reinterpret_cast<std::vector<std::shared_ptr<Maths::QuadTreePoint>>*>(vector)->at(index).get());
+    try
+    {
+      *outputPoint = reinterpret_cast<NrtQuadTreePoint>(reinterpret_cast<std::vector<std::shared_ptr<Maths::QuadTreePoint>>*>(vector)->at(index).get());
+    }
+    catch (const std::out_of_range)
+    { // todo: handle error message
```
Comment (RubyNova) [inline src/NovelRT.Interop/Maths/NrtQuadTree.cpp:164]: Was this TODO handled or are we leaving it for this PR?

**Review by RubyNova — CHANGES_REQUESTED** [review]
Looking pretty good overall! I have one definite change that needs doing but the rest are TODO related questions.

**Review by RubyNova — APPROVED** [review]
Based on a discussion on Discord, this PR will cause a new ticket to be created for the missing error messages. Otherwise, LGTM!

**Pull request merged**